### PR TITLE
Integrate FAM pipeline into robottelo

### PR DESCRIPTION
**Pull request by Griffin-Sullivan**
### Problem Statement
Current FAM misc pipeline has been broken for a couple releases. This is due to some version changes, but also we were not testing things as accurately as possible. For instance, we were cloning upstream and running the develop branch.

### Solution
Integrate the FAM pipeline into robottelo. This now runs everything on the satellite against itself, so we know that the collection being packaged in Satellite is working correctly.

### Note
The expectation is not that this will pass 100%. Getting all of the modules and roles to pass will take individual setup that needs to be done in robottelo, AAP, or upstream in the test playbook itself. For now I would like to see that the tests are executing as expected and the results show helpful output on failure (ie: the actual ansible tasking logs). After merging, we will need to evaluate as teams how to remediate the issues based on component and team.

### CODE REVIEW

**Comment by Griffin-Sullivan** [comment]
trigger: test-robottelo
pytest: tests/foreman/sys/test_fam.py::test_positive_run_modules_and_roles

**Comment by Griffin-Sullivan** [comment]
Not a terrible start (30/86 failed). You can see though that the failure output is not very helpful since it doesn't give you the full ansible play logs. I'm gonna research if there's a way to force Jenkins or pytest to show everything.

File: tests/foreman/sys/test_fam.py [hunk 1]
```diff
@@ -32,6 +41,48 @@ def sync_roles(target_sat):
...
+    )
+    module_target_sat.execute(
+        f'''sed -i 's/rhsm_username:.*/rhsm_username: "{settings.subscription.rhn_username}"/g' {config_file}'''
+    )
+    module_target_sat.execute(
+        f'''sed -i 's|subscription_manifest_path:.*|subscription_manifest_path: "data/{module_sca_manifest.name}.zip"|g' {config_file}'''
+    )
+    module_target_sat.execute(
+        f'''sed -i 's/rhsm_password:.*/rhsm_password: "{settings.subscription.rhn_password}"/g' {config_file}'''
+    )
```
Comment (evgeni) [inline tests/foreman/sys/test_fam.py:77]: given the amount of changes we need to perform here, would it be easier/cleaner if we'd deploy this file (`server.yml`) from a template and completely ignore the example we ship in the package?

File: tests/foreman/sys/test_fam.py [hunk 2]
```diff
@@ -32,6 +41,48 @@ def sync_roles(target_sat):
...
+
+    # Handle color output formatting
+    module_target_sat.execute('export ANSIBLE_NOCOLOR=True && export PY_COLORS=0')
```
Comment (evgeni) [inline tests/foreman/sys/test_fam.py:80]: does `.execute()` always start a new ssh/shell process? if so, those `export`s won't do anything

Comment (Griffin-Sullivan) [reply tests/foreman/sys/test_fam.py:80]: I'm actually not sure. Good question for @JacobCallahan

File: tests/foreman/sys/test_fam.py (hunk 1 above)
Comment (evgeni) [reply tests/foreman/sys/test_fam.py:77]: No idea ;)
I was hoping there is just a way in robotello?

**Comment by Griffin-Sullivan** [comment]
trigger: test-robottelo
pytest: tests/foreman/sys/test_fam.py::test_positive_run_modules_and_roles

**Comment by Griffin-Sullivan** [comment]
Reran cause I realized I needed to add some more test_playbooks to run. This is now running the right amount of tests.

File: tests/foreman/sys/test_fam.py (hunk 2 above)
Comment (evgeni) [reply tests/foreman/sys/test_fam.py:79]: Looking at https://github.com/SatelliteQE/broker/blob/master/broker/hosts.py, it seems the ssh session is kept alive (if possible).

I'd probably still move that to the actual test invocation below.

File: tests/foreman/sys/test_fam.py (hunk 1 above)
Comment (evgeni) [reply tests/foreman/sys/test_fam.py:76]: I've found https://github.com/SatelliteQE/robottelo/blob/97d9117d74835665cad06b59d39ab5802a63b0ec/robottelo/hosts.py#L881-L892

Comment (evgeni) [reply tests/foreman/sys/test_fam.py:76]: or https://github.com/SatelliteQE/robottelo/blob/97d9117d74835665cad06b59d39ab5802a63b0ec/robottelo/hosts.py#L1003-L1018

File: tests/foreman/sys/test_fam.py [hunk 3]
```diff
@@ -32,6 +40,48 @@ def sync_roles(target_sat):
...
+    module_target_sat.execute(
+        f"sed -i 's/pytest -vv/pytest -v --color=no /g' {FAM_ROOT_DIR}/Makefile"
```
Comment (evgeni) [inline tests/foreman/sys/test_fam.py:81]: you want `-vv` here, which will make it output all ansible output (as you wanted)

**Comment by Griffin-Sullivan** [comment]
trigger: test-robottelo
pytest: tests/foreman/sys/test_fam.py::test_positive_run_modules_and_roles

File: tests/foreman/sys/test_fam.py [hunk 4]
```diff
@@ -32,6 +40,47 @@ def sync_roles(target_sat):
...
+    module_target_sat.execute(
+        f'''sed -i 's|subscription_manifest_path:.*|subscription_manifest_path: "data/{module_sca_manifest.name}.zip"|g' {config_file}'''
+    )
+    module_target_sat.execute(
+        f'''sed -i 's/rhsm_password:.*/rhsm_password: "{settings.subscription.rhn_password}"/g' {config_file}'''
+    )
+
+    # Handle color output formatting
+    module_target_sat.execute(
+        f"sed -i 's/pytest -vv/pytest -vv --color=no /g' {FAM_ROOT_DIR}/Makefile"
```
Comment (evgeni) [inline tests/foreman/sys/test_fam.py:80]: setting `PY_COLORS=0` below should be sufficient to force colors off. Sure this is required?

File: tests/foreman/sys/test_fam.py [hunk 5]
```diff
@@ -71,3 +120,20 @@ def test_positive_import_run_roles(sync_roles, target_sat):
...
+def test_positive_run_modules_and_roles(module_target_sat, setup_fam, ansible_module):
+    """Run all FAM modules and roles on the Satellite
+
+    :id: b595756f-627c-44ea-b738-aa17ff5b1d39
+
+    :expectedresults: All modules and roles run successfully
+    """
+    # Execute test_playbook
+    result = module_target_sat.execute(
+        'export ANSIBLE_NOCOLOR=True && export PY_COLORS=0 && '
```
Comment (evgeni) [inline tests/foreman/sys/test_fam.py:136]: Both [Ansible](https://docs.ansible.com/ansible/latest/reference_appendices/config.html#ansible-nocolor) and [Pytest](https://docs.pytest.org/en/8.0.x/reference/reference.html#envvar-NO_COLOR) support [`NO_COLOR`](https://no-color.org/) so it should be sufficient to set `NO_COLOR=True`

File: tests/foreman/sys/test_fam.py [hunk 6]
```diff
@@ -71,3 +120,20 @@ def test_positive_import_run_roles(sync_roles, target_sat):
...
+        f'. ~/localenv/bin/activate && cd {FAM_ROOT_DIR} && make livetest_{ansible_module}'
+    )
+    assert 'PASSED' in result.stdout
```
Comment (evgeni) [inline tests/foreman/sys/test_fam.py:139]: I think you want to also assert exit code being zero, just to be sure?

**Comment by evgeni** [comment]
I think I know why we're missing the "real" output in the test results.

The pytest that is running the "outer" tests is not called with `-vv`:
```
18:36:18  + pytest tests/foreman/sys/test_fam.py::test_positive_run_modules_and_roles --external-logging --junit-xml=sat-results.xml -o junit_suite_name=sat-result
```

File: tests/foreman/sys/test_fam.py (hunk 4 above)
Comment (Griffin-Sullivan) [reply tests/foreman/sys/test_fam.py:80]: ok and I'm not sure. I never made those changes originally

**Comment by Griffin-Sullivan** [comment]
> I think I know why we're missing the "real" output in the test results.
> 
> The pytest that is running the "outer" tests is not called with `-vv`:
> 
> ```
> 18:36:18  + pytest tests/foreman/sys/test_fam.py::test_positive_run_modules_and_roles --external-logging --junit-xml=sat-results.xml -o junit_suite_name=sat-result
> ```

Interesting. I'll try changing it. Let me update everything and we'll rerun.

**Comment by Griffin-Sullivan** [comment]
trigger: test-robottelo
pytest: -vv tests/foreman/sys/test_fam.py::test_positive_run_modules_and_roles

**Comment by Satellite-QE** [comment]
**PRT Result**
```
Build Number: 5721
Build Status: UNSTABLE
PRT Comment: pytest -vv tests/foreman/sys/test_fam.py::test_positive_run_modules_and_roles
Test Result : =========== 41 failed, 69 passed, 137 warnings in 3850.63s (1:04:10) ===========
```

**Comment by evgeni** [comment]
Well, at least it didn't eat up the output now

File: tests/foreman/sys/test_fam.py (hunk 1 above)
Comment (Griffin-Sullivan) [reply tests/foreman/sys/test_fam.py:76]: We would still have to replace the hostname and manifest name during the test setup. For now I'd leave this as is even though it's ugly and we can change in the future.

**Review by evgeni — APPROVED** [review]

File: tests/foreman/sys/test_fam.py [hunk 7]
```diff
@@ -32,6 +40,42 @@ def sync_roles(target_sat):
...
+@pytest.fixture(scope='module')
+def setup_fam(module_target_sat, module_sca_manifest):
+    # Execute AAP WF for FAM setup
+    Broker().execute(workflow='fam-test-setup', source_vm=module_target_sat.name)
+
+    # Edit Makefile to not try to rebuild the collection when tests run
+    module_target_sat.execute(f"sed -i '/^live/ s/$(MANIFEST)//' {FAM_ROOT_DIR}/Makefile")
+
+    # Upload manifest to test playbooks directory
+    module_target_sat.put(f'{module_sca_manifest.path}', f'{module_sca_manifest.name}')
```
Comment (JacobCallahan) [inline tests/foreman/sys/test_fam.py:52]: why the f-strings here?

**Review by JacobCallahan — CHANGES_REQUESTED** [review]
great addition! just one minor thing

**Review by JacobCallahan — APPROVED** [review]

**Review by ogajduse — APPROVED** [review]
Co-reviewed with @jameerpathan111 
LGTM, let's see how it behaves in the production.

**Review by jameerpathan111 — APPROVED** [review]

**Pull request merged**